sdk/include/ndk/haltypes.h

```diff
@@ -39,61 +162,1375 @@ typedef enum _FIRMWARE_REENTRY
+            UCHAR Device;
+            UCHAR Function;
+            UCHAR Reserved;
+        } DeviceRid;
+        struct {
+            UCHAR Bus;
+            UCHAR Device;
+            UCHAR Function;
+            UCHAR Reserved;
+        } BridgeRid;
+        struct {
+            UCHAR SecondaryBus;
+            UCHAR SubordinateBus;
+        } MultipleBridges;
+    } DUMMYSTRUCTNAME;
+} PCI_BUSMASTER_DESCRIPTOR, *PPCI_BUSMASTER_DESCRIPTOR;
+
+typedef
+NTSTATUS
+(NTAPI *pHalAddInterruptRemapping)(
+    _In_ ULONG BusNumber,
+    _In_ ULONG SlotNumber,
+    _In_ PPCI_BUSMASTER_DESCRIPTOR BusMasterDescriptor,
+    _In_range_(0, 3) UCHAR PhantomBits,
+    _Inout_updates_(VectorCount) PINTERRUPT_VECTOR_DATA VectorData,
+    _In_ ULONG VectorCount
+);
+
+typedef
+VOID
+(NTAPI *pHalRemoveInterruptRemapping)(
+    _In_ ULONG BusNumber,
+    _In_ ULONG SlotNumber,
+    _In_ PPCI_BUSMASTER_DESCRIPTOR BusMasterDescriptor,
+    _In_range_(0, 3) UCHAR PhantomBits,
+    _Inout_updates_(VectorCount) PINTERRUPT_VECTOR_DATA VectorData,
+    _In_ ULONG VectorCount
+);
+
+typedef
+VOID
+(NTAPI *pHalSaveAndDisableHvEnlightenment)(
+    VOID
+);
+
+typedef
+VOID
+(NTAPI *pHalRestoreHvEnlightenment)(
+    VOID
+);
+
+typedef
+VOID
+(NTAPI *pHalFlushIoBuffersExternalCache)(
+    _In_ PMDL  Mdl,
+    _In_ BOOLEAN  ReadOperation
+);
+
+typedef
+VOID
+(NTAPI *pHalFlushIoRectangleExternalCache)(
+    _In_ PMDL Mdl,
+    _In_ ULONG StartOffset,
+    _In_ ULONG Width,
+    _In_ ULONG Height,
+    _In_ ULONG Stride,
+    _In_ BOOLEAN ReadOperation
+);
+
+typedef
+VOID
+(NTAPI *pHalFlushExternalCache)(
+    _In_ BOOLEAN Invalidate
+);
+
+typedef
+VOID
+(NTAPI *pHalFlushAndInvalidatePageExternalCache)(
+    _In_ PHYSICAL_ADDRESS PhysicalAddress
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalPciEarlyRestore)(
+    _In_ SYSTEM_POWER_STATE SleepState
+);
+
+typedef
+VOID
+(NTAPI *pHalPciLateRestore)(
+    VOID
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalGetProcessorId)(
+    _In_ ULONG ProcessorIndex,
+    _Out_ ULONG *Identifier
+);
+
+typedef struct _HAL_PMC_COUNTERS *PMC_HANDLE;
+
 typedef
-PBUS_HANDLER
-(FASTCALL *pHalHandlerForConfigSpace)(
-    _In_ BUS_DATA_TYPE ConfigSpace,
-    _In_ ULONG BusNumber
+NTSTATUS
+(NTAPI *pHalAllocatePmcCounterSet)(
+    _In_ ULONG ProcessorIndex,
+    _In_reads_(SourceCount) KPROFILE_SOURCE *SourceList,
+    _In_ ULONG SourceCount,
+    _Out_ PMC_HANDLE *Handle
+);
+
+typedef
+VOID
+(NTAPI *pHalFreePmcCounterSet)(
+    _In_ PMC_HANDLE Handle
+);
+
+typedef
+VOID
+(NTAPI *pHalCollectPmcCounters)(
+    _In_ PMC_HANDLE Handle,
+    _Out_ PULONG64 Data
+);
+
+typedef
+ULONGLONG
+(NTAPI *pHalTimerQueryCycleCounter)(
+    _Out_opt_ PULONGLONG CycleCounterFrequency
+);
+
+typedef
+VOID
+(NTAPI *pHalGetNextTickDuration)(
+    PKPRCB Unknown1,
+    BOOLEAN Unknown2,
+    ULONG Unknown3,
+    ULONG64 Unknown4,
+    PULONGLONG Unknown5
 );
 
 typedef
 NTSTATUS
-(NTAPI *PINSTALL_BUS_HANDLER)(
-    _In_ PBUS_HANDLER Bus
+NTAPI
+PROCESSOR_HALT_ROUTINE(
```

Comment:
@archeYR Is it possible to move this `PROCESSOR_HALT_ROUTINE` definition to the wdm section of _**xdk**_/haltypes.h ?

Reply:
This definition is used by the HAL function below. Also `xdk/haltypes.h` doesn't have a `_WDMDDK_` section.

Reply:
Then it probably needs such a section to be added ^^